src/app/fdctl/run/tiles/fd_dedup.c

```diff
@@ -114,13 +120,65 @@ after_frag( void *             _ctx,
             ulong *            opt_tsorig,
             int   *            opt_filter,
             fd_mux_context_t * mux ) {
-  (void)in_idx;
   (void)seq;
   (void)opt_tsorig;
   (void)mux;
 
   fd_dedup_ctx_t * ctx = (fd_dedup_ctx_t *)_ctx;
 
+  if( FD_UNLIKELY( in_idx == GOSSIP_IN_IDX ) ) {
+    /* Transactions coming in from the gossip link are not parsed.
+
+       We'll need to parse it so it's ready for downstream consumers.
+       Equally importantly, we need to parse to extract the signature for dedup.
+       Just parse it right into the output dcache. */
+
+    ulong payload_sz = *opt_sz;
+    ulong fd_txn_off = fd_ulong_align_up( payload_sz, 2UL );
+
+    /* Ensure sufficient trailing space for parsing. */
+    long txnt_maxsz = (long)FD_TPU_DCACHE_MTU -
+                      (long)fd_txn_off -
+                      (long)sizeof(ushort);
+    if( FD_UNLIKELY( txnt_maxsz<(long)FD_TXN_MAX_SZ ) ) {
+      FD_LOG_ERR(( "got malformed txn (sz %lu) insufficient space left in dcache for fd_txn_t", payload_sz ));
+    }
+
+    uchar const * dcache_entry = fd_chunk_to_laddr_const( ctx->out_mem, ctx->out_chunk );
+    fd_txn_t    * fd_txn = (fd_txn_t *)((ulong)dcache_entry + fd_txn_off);
+    ulong txn_t_sz = fd_txn_parse( dcache_entry, payload_sz, fd_txn, NULL );
+    if( FD_UNLIKELY( !txn_t_sz ) ) {
+      FD_LOG_WARNING(( "fd_txn_parse failed for gossiped vote" ));
+      *opt_filter = 1;
+      return;
+    }
+
+    /* Write payload_sz into trailer.
+       fd_txn_parse always returns a multiple of 2 so this sz is
+       correctly aligned. */
+    ushort * payload_sz_p = (ushort *)( (ulong)fd_txn + txn_t_sz );
+    *payload_sz_p = (ushort)payload_sz;
+
+    /* End of parsed message. */
+
+    ulong new_sz = ( (ulong)payload_sz_p + sizeof(ushort) ) - (ulong)dcache_entry;
+    if( FD_UNLIKELY( new_sz>FD_TPU_DCACHE_MTU ) ) {
+      FD_LOG_CRIT(( "memory corruption detected (txn_sz=%lu txn_t_sz=%lu)",
+                    payload_sz, txn_t_sz ));
+    }
```

Comment:
Not really any new constraint as far as I can tell.  I took this from fd_verify.c and it seems to be an attempt to re-check an invariant that touches on several variables after fd_txn_parse().  Parsing of untrusted input is somewhat tricky after all, so I can see the point of being paranoid post parsing.  That being said, the check is more of a bug trap rather than a security measure.